### ISSUE TITLE
OutOfMemoryException in variable substitution

### ISSUE THREAD

**Issue by ammaraltahan** [DbUp/DbUp#189]
I have a large file (data inserts) that has one variable at the top of the file (i.e.Using [$Database_Name$]) and it errors out with this exception.

Is this a limitation for Regex? or the regex is too greedy? may be worth looking into.

System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was
 thrown.
   at System.Text.RegularExpressions.RegexRunner.DoubleTrack()
   at System.Text.RegularExpressions.RegexRunner.EnsureStorage()
   at System.Text.RegularExpressions.RegexInterpreter.Goto(Int32 newpos)
   at System.Text.RegularExpressions.RegexInterpreter.Go()
   at System.Text.RegularExpressions.RegexRunner.Scan(Regex regex, String text,
Int32 textbeg, Int32 textend, Int32 textstart, Int32 prevlen, Boolean quick, Tim
eSpan timeout)
   at System.Text.RegularExpressions.Regex.Run(Boolean quick, Int32 prevlen, Str
ing input, Int32 beginning, Int32 length, Int32 startat)
   at System.Text.RegularExpressions.Regex.Match(String input, Int32 startat)
   at System.Text.RegularExpressions.Regex.Match(String input)
   at DbUp.Engine.Preprocessors.VariableSubstitutionPreprocessor.Process(String

**Comment by omnibs** [comment]
What's the size of the file? I'll try to reproduce the issue and see what can be done.

**Comment by ammaraltahan** [comment]
46.5 Mbs !! :)

**Comment by omnibs** [comment]
That's not a lot! How much ram do you have? Are you running as a console app or embedded in an ASP.NET application?

**Comment by ammaraltahan** [comment]
I have 16g RAM and yes I'm running as console app.

**Comment by omnibs** [comment]
Just tried adding a 52mb script with lots of comments and random `select *` lines to the sample app in the repo. Ran just fine. 

I'd guess there's something specific to your script that's making the regex go crazy, like [this Atom regex bug](http://davidvgalbraith.com/how-i-fixed-atom/) that made the editor hang whenever enter key was pressed at the end of a very specific line of Go code.

My suspect is this huge regex in `VariableSubstitutionPreprocessor` to prevent variables from being substituted inside comments:

```c#
new Regex(@"/\*(?>(?:(?!\*/|/\*).)*)(?>(?:/\*(?>(?:(?!\*/|/\*).)*)\*/(?>(?:(?!\*/|/\*).)*))*).*?\*/|--.*?\r?[\n]", RegexOptions.Singleline);
```

If you can't share your script file for me to have a look, could your try running it on DbUp, from source, changing the [`VariableSubstitutionPreprocessor.cs`](https://github.com/DbUp/DbUp/blob/master/src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs) to the code in [this Gist](https://gist.github.com/omnibs/21b6685e88ab49f92583578751c07957)?

**Comment by ammaraltahan** [comment]
Thanks for looking into it. I don't know if I'm allowed to share that sql
script, but to be safe I'll do your suggestion and let you know.

On Wed, Apr 27, 2016 at 3:16 PM, Juliano Solanho <notifications@github.com>
wrote:

> Just tried adding a 52mb script with lots of comments and random select *
> lines to the sample app in the repo. Ran just fine.
>
> I'd guess there's something specific to your script that's making the
> regex go crazy, like this Atom regex bug
> <http://davidvgalbraith.com/how-i-fixed-atom/> that made the editor hang
> whenever enter key was pressed at the end of a very specific line of Go
> code.
>
> My suspect is this huge regex in VariableSubstitutionPreprocessor to
> prevent variables from being substituted inside comments:
>
> new Regex(@"/\*(?>(?:(?!\*/|/\*).)*)(?>(?:/\*(?>(?:(?!\*/|/\*).)*)\*/(?>(?:(?!\*/|/\*).)*))*).*?\*/|--.*?\r?[\n]", RegexOptions.Singleline);
>
> If you can't share your script file for me to have a look, could your try
> running it on DbUp, from source, changing the
> VariableSubstitutionPreprocessor.cs
> <https://github.com/DbUp/DbUp/blob/master/src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs>
> to the code in this Gist
> <https://gist.github.com/omnibs/21b6685e88ab49f92583578751c07957>?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/DbUp/DbUp/issues/189#issuecomment-215198244>
>



-- 
Ammar Altahan

**Comment by ammaraltahan** [comment]
the script ran fine with code you shared in your Gist. it does blow up in the first Regex.Match.
ie. Match commentMatch = regexObj.Match(contents);

**Comment by ammaraltahan** [comment]
what does the first regex do anyways??

**Comment by ammaraltahan** [comment]
One more clue @omnibs , I have auto generated the sql file from Sql Management studio (right-click DB then tasks then Generate Script then from Advanced options select Data Only) 
If you have a large table, you can generate an insert sql file and see if you can reproduce it. here I can easily reproduce. Right now I have copied your code into a custom preprocessor and configured it to work with that. But it would be nice if we find the root cause.

**Comment by omnibs** [comment]
Thanks for the effort in helping identify this @ammaraltahan =]

I did try generating scripts for a big table (75mb .sql file), I added a `use $somevar$` to the top and ran it just fine with the current regex code.

I then tried creating a 52mb file made entirely of `use $somevar$`, but still it ran just fine.

My guess is you might have html/javascript/xml inside your inserts and those might have some combination of `/*` and `*/` that's making that regex backtrack to infinity.

I think the fact that the OutOfMemory is happening in that specific Regex is reason enough for us to try a different approach there.

**Comment by ammaraltahan** [comment]
Thanks for the help and the efforts. I'll scan the SQL file for elements
you mentioned.

**Comment by omnibs** [comment]
I've reimplemented the IgnoreComments functionality without Regex in a branch here: https://github.com/omnibs/DbUp/tree/remove-comment-regex

Do you mind testing it with your script?

If you'd like to test it as a custom preprocessor, you'll need these two files:

[VariableSubstitutionPreprocessor.cs](https://github.com/omnibs/DbUp/blob/remove-comment-regex/src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs)
[SqlCommentRangeFinder.cs](https://github.com/omnibs/DbUp/blob/remove-comment-regex/src/DbUp/Helpers/SqlCommentRangeFinder.cs)

**Comment by ammaraltahan** [comment]
Sure thing. I'll test it tomorrow morning.

**Comment by ammaraltahan** [comment]
Hey @omnibs I've tested the branch, and it's working fine with the problematic script.

**Comment by omnibs** [comment]
Thanks, I'll make a pull request out of that =]

**Comment by ammaraltahan** [comment]
I've done some more testing on master branch with the problematic SQL
script, I found that '/' is breaking the regex, we have strings that have
forward slash in them. I'll try a regex tool on it and let you know.

**Comment by ammaraltahan** [comment]
hey @omnibs , I've done more thorough testing, it is confirmed that the "/" breaks the regex in a large file like that one I have.

**Comment by omnibs** [comment]
Thanks @ammaraltahan! 

I'm sort of decided into removing the regex anyway, I think it's faster and easier to reason with the parser. 

There a discussion at PR #190, we're thinking about reusing another SQL syntax interpreter we have in the codebase that would make comment identification even more robust.

I'll close here once we decide and merge the fix.

**Comment by InvincibleDRT** [comment]
do you have any idea on how to run a 380MB sql file using dbup. I am getting Out of Memory exception. I could try splitting up that file, but i would like to know any other alternatives before doing that

**Comment by droyad** [comment]
Fixed in #190
